### PR TITLE
Geometry chunks serialization

### PR DESCRIPTION
**Pull request by AntoineDao**
Hi guys! I'm digging `pyspecklev2`! I mean, `specklepy`!!

I was playing around with integrating some custom object models through the Python client and was testing decoding a complex Brep to figure out how to reproduce it from some other object (you know... a honeybee model :bee: ).

Anyways, while I was messing with that I realised that there was no methods to deal with data chunks. This PR is meant to resolve this to the best of my understanding. Happy for any feedback/changes if it helps get this one through so I can go back to messing with honeybee schemas and speckle objects :raised_hands:

### CODE REVIEW

**Comment by izzylys** [comment]
Hey Antoine! Lovely to see you around again 😊

With the chunking, the way it works right now is that a chunkable list of objects will be serialised as a list of serialised objects eg:

```json
{
  "speckle_type": "Speckle.Core.Models.DataChunk",
  "data": [
    {
      "x": 1.0,
      "y": 2.0,
      "z": 0.0,
      "speckle_type": "Objects.Geometry.Point"
    },
    {
      "x": 3.0,
      "y": 2.0,
      "z": 1.0,
      "speckle_type": "Objects.Geometry.Point"
    }
  ]
}
```

Of course, chunkable lists of primitives eg doubles will be chunked as just lists of doubles.

This chunking / dechunking is handled in the serialisation / deserialisation process. May I ask why you want more individual control of how the chunks work? In theory, you shouldn't have to worry about this at all as long as you've defined any custom chunkable attributes on the objects to the serialiser can pick this up and deal with it. Whether you're receiving with `specklepy` or `speckle sharp`, both should handle this for you on `operations.send` and `operations.receive`

Side note: regarding tests, you'll need to compare objects by computing their ids with (`get_id()`) instead of comparing a deserialised object with an original object since the original one won't have an id so the two won't be a match. However, the `test_decode_arbitrary_object_chunks` still seems to be failing for me even with this fix.

Side side note: serialisation is about to change with #118 and will be faster and better so look out for that! 🚀

**Comment by AntoineDao** [comment]
Hi @izzylys :wave:

Sorry for the late reply on my end, other stuff picked my attention :sweat_smile: I might have misunderstood the existing API exposed by the python client.

## Deserialization from SpeckleServer

I encountered this issue when trying to fetch [an existing Brep object from the public speckle server](https://speckle.xyz/streams/87a17f4280/objects/ece866af9e3655f59936b08df3fa47a9) to better understood what was in it. When I did this I noticed the object was missing all of the `chunkable` attributes and they were kept in their `list` format instead of dict objects. I might not have been using the right functions...

Here is the code I used to execute this:

```python
from specklepy.api.credentials import get_default_account, get_local_accounts
from specklepy.api.client import SpeckleClient
from specklepy.objects.geometry import Brep
from specklepy.serialization.base_object_serializer import BaseObjectSerializer
from specklepy.transports.server import ServerTransport
from specklepy.transports.memory import MemoryTransport

token = '<my-speckle-token>'

stream_id = '87a17f4280'
object_id = 'ece866af9e3655f59936b08df3fa47a9'

all_accounts = get_local_accounts()  # get back a list
account = get_default_account()

client = SpeckleClient(host="speckle.xyz")
client.authenticate(token)

transport = ServerTransport(client=client, stream_id=stream_id)
cache = MemoryTransport()

transport.copy_object_and_children(id=object_id, target_transport=cache)


serializer = BaseObjectSerializer(read_transport=cache)

received_base = client.object.get(stream_id=stream_id, object_id=object_id)

brep: Brep = serializer.recompose_base(received_base.dict())

print(brep)
```

**Comment by izzylys** [comment]
hey @AntoineDao !

I think there's just been a bit of a misunderstanding here 😅 

so, if you're just receiving an object then this is the simplest way to go about it assuming you have you accounts locally thanks to the Speckle Manager (which it looks like you do since you're using the credentials stuff successfully). you should just be able to run this as is

```py
from specklepy.api import operations
from specklepy.api.credentials import StreamWrapper
from specklepy.transports.server import ServerTransport

object_url = "https://speckle.xyz/streams/87a17f4280/objects/ece866af9e3655f59936b08df3fa47a9"

# this is a nice helper (like we have in `sharp`) for extracting details from a url
wrapper = StreamWrapper(object_url)

# you can use the wrapper to get an authenticated client and server transport
server_transport = wrapper.get_transport()

# the `operations.send` and `operations.receive` are what you use to handle complex objects
received_brep = operations.receive(wrapper.object_id, server_transport)

print(received_brep)
# Brep(id: ece866af9e3655f59936b08df3fa47a9, speckle_type: Objects.Geometry.Brep, totalChildrenCount: 11)
```

you can find more details and other examples in the [docs](https://speckle.guide/dev/py-examples.html)

basically, you as a user just sending and receiving should not have to interact with the serialiser at all! the `client.object.get()` is just doing the GraphQL API call to get the object from the server without doing anything with it for you. if you have a complex object, it is definitely recommended to use `operations.send` and `operations.receive`.

hope that clears things up! as i mentioned, you can find plenty more info and examples [in the docs at speckle.guide](https://speckle.guide/dev/py-examples.html) but feel free to ask if anything is still unclear!

**Comment by AntoineDao** [comment]
Nice! That's definitely a cleaner way of making that API call :sweat_smile: That being said two problems remain on my end running this script:

1. I'm a pretentious person so I like to develop in a Linux environment (and for a Linux environment) so the Speckle client can't use credentials stored by SpeckleManager because it doesn't work on Linux :smile:
2. Using your example code on the latest `main` branch still doesn't deserialize the chunked attributes into the correct Base objects

I used the python debugger in VSCode to check the properties of the Brep and as you can see from the image below that the chunked values are not parsed into their corresponding keys:
image

Here's the code I used so you can check I've not got some weird config going on :sweat_smile: 

```python
from specklepy.api import operations
from specklepy.api.credentials import StreamWrapper

object_url = "https://speckle.xyz/streams/87a17f4280/objects/ece866af9e3655f59936b08df3fa47a9"

# this is a nice helper (like we have in `sharp`) for extracting details from a url
wrapper = StreamWrapper(object_url)

client = wrapper.get_client()
token = '<token>'
client.authenticate(token)


# you can use the wrapper to get an authenticated client and server transport
server_transport = wrapper.get_transport()

# the `operations.send` and `operations.receive` are what you use to handle complex objects
received_brep = operations.receive(wrapper.object_id, server_transport)

print("Received Brep")
print(received_brep)
# Brep(id: ece866af9e3655f59936b08df3fa47a9, speckle_type: Objects.Geometry.Brep, totalChildrenCount: 11)
print()

print("Curve 2D parsed and chunked values")
print(received_brep.Curve2D)
# None
print(received_brep.Curve2DValues)
# [10, 4, -24, etc...]
print()

print("Curve 3D parsed and chunked values")
print(received_brep.Curve3D)
# None
print(received_brep.Curve3DValues)
# [10, 4, 41, etc...]
print()

print("Surfaces parsed and chunked values")
print(received_brep.Surfaces)
# None
print(received_brep.SurfacesValue)
# [35, 1, 1, etc...]
print()

print("Vertices parsed and chunked values")
print(received_brep.Vertices)
# None
print(received_brep.VerticesValue)
# [3, 41.8, -11.3, etc...]
```

Is it maybe worth having a quick chat to figure this one out when you have a minute? (Doesn't have to be today of course!)

**Comment by izzylys** [comment]
heya!

for 1) that's totally fine, you can use the more verbose way of getting server transport as you were doing before. you can also add accounts manually as shown [here](https://speckle.guide/user/manager.html#adding-accounts-manually) which will also get picked up by py and the other connectors (I have had other linux users attest to this!).

for 2) it looks like your brep object doesn't actually have any `Curve2D` and `Curve3D` values if you have a look at it online. it seems like this is more of a problem with your send than the receive.

https://speckle.xyz/streams/87a17f4280/objects/ece866af9e3655f59936b08df3fa47a9
image

however, you do have `Curve2DValues` and `Curve3DValues` which are chunked lists and these are coming through correctly from your screenshot (ie as one single long list of values)

happy to hop on a call if you'd like tho!

**Comment by AntoineDao** [comment]
Hi @izzylys :wave: 

I reworked this one from scratch and did a force push to the same branch. Here's what's been done:
1. Added `to_list` and `from_list` for geometry objects that have it in C# and add tests
2. Add some methods to the `DataChunk` class to facilitate encoding and decoding lists of Base objects that have a `to_list` and `from_list` method. Also added a test for that one
3. Added an `ignore_serialize` attribute to the `Base` `__init_subclass__` function so users can define custom "non serializable" attributes. This is meant to be the equivalent of what has a `JsonIgnore` decorator in the C# library
4. (Finally!) rewrite the `Brep` object to expose geometry attributes that are stored as lists of floats

File: specklepy/objects/base.py [hunk 1]
```diff
@@ -403,3 +410,45 @@ class DataChunk(Base, speckle_type="Speckle.Core.Models.DataChunk"):
...
+            chunk_end = int(chunk_start + chunk_length)
+            chunk_data = data[chunk_start:chunk_end]
+            decoded_data = decoder(chunk_data)
+            if isinstance(decoded_data, Base):
+                decoded_data.id = decoded_data.get_id()
+            unchunked_data.append(decoded_data)
+            index = chunk_end
+        return unchunked_data
+
+    def decode(self, decoder: Callable[[List[Any]], Any]):
```
Comment (izzylys) [inline specklepy/objects/base.py:448]: missing a ref to `Callable` at the top of this file ;)

File: specklepy/objects/base.py [hunk 2]
```diff
@@ -317,8 +324,8 @@ def units(self, value: str):
 
     def get_member_names(self) -> List[str]:
         """Get all of the property names on this object, dynamic or not"""
-        # attrs = set(self.__dict__.keys())
-        attr_dir = list(set(dir(self)) - REMOVE_FROM_DIR)
+        ignored_attributes = REMOVE_FROM_DIR.union(self._serialize_ignore)
```
Comment (izzylys) [inline specklepy/objects/base.py:327]: the users should actually be able to see these attrs. they are interesting to the user and ignoring them is _only_ for the benefit of serialisation and deserialisation. instead, we should do this in the deserialiser when traversing through the attributes and ignore it if it exists in `_serialize_ignore`

File: specklepy/objects/base.py [hunk 3]
```diff
@@ -403,3 +410,45 @@ class DataChunk(Base, speckle_type="Speckle.Core.Models.DataChunk"):
 
     def __init__(self) -> None:
         self.data = []
+
+    @classmethod
```
Comment (izzylys) [inline specklepy/objects/base.py:414]: I don't think the `DataChunk` should contain these methods. the `DataChunk` is just a dumb class whose purpose is to hold data. the serialisation / deserialisation is what deals with chunking and dechunking of lists without any relation to the data models. the `DataChunk` shouldn't deal with creating / deserialising objects itself.

File: specklepy/objects/geometry.py [hunk 4]
```diff
@@ -313,45 +563,146 @@ def _edge(self):
...
+    def _inject_self_into_children(self, children: Optional[List[Base]]) -> List[Base]:
+        if children is None:
+            return children
+
+        for child in children:
+            child._Brep = self
+        return children
+
+    @property
+    def Edges(self) -> List[BrepEdge]:
```
Comment (izzylys) [inline specklepy/objects/geometry.py:632]: as mentioned in the `DataChunk` comments, chunking should not be exposed to the users at all. similarly, it should be the `Edges`, `Loops`, `Trims`, etc that are defined on the class and the `{abc}Values` that have getters and setters since the `Values` lists are just for the benefit of serialisation / deserialisation.

eg `Curve3D` is defined on the class and is what the user interacts with when building and inspecting a brep
`Curve3DValues` has:
- a getter that returns `Curve3D` as a flat list of numbers (which the serialiser will chunk / dechunk as it does any other list)
- a setter that takes a flat list, decodes it to a list of curves, and sets that to `Curve3D`

**Review by izzylys — CHANGES_REQUESTED** [review]
Hey @AntoineDao,

Thanks so much for your work on this! Also thanks for the chat so we could clear up the confusion I had re this 😄. Appreciate you endlessly and also owe you a beer for catching this mismatch between sharp and py re breps 🍻 

This is almost perfect and ready to merge other than a bit of a mixup in the `Brep` model.

Basically, chunking and the `{abc}Values` lists should not be seen by the user since it's not important to them. Like in `sharp`, we want the user to interact with `Breps` through the `Curves`, `Faces`, `Edges`, `Loops`, etc and have the `{abc}Values` lists only for serialisation / deserialisation. I've added some comments in the code. Let me know what ya think / if you disagree / if I'm being unclear!

Also feel free to tell me you're fed up with this and want me to make the changes if you want 😅

File: specklepy/objects/base.py (hunk 2 above)
Comment (AntoineDao) [reply specklepy/objects/base.py]: Resolved in 3005e42

File: specklepy/objects/base.py (hunk 3 above)
Comment (AntoineDao) [reply specklepy/objects/base.py]: Resolved in 0270219

File: specklepy/objects/geometry.py (hunk 4 above)
Comment (AntoineDao) [reply specklepy/objects/geometry.py:633]: I'm torn about this change. I see where you're coming from with regards to exposing the parameters users interact with and obfuscating the stuff they shouldn't be worrying about. I also appreciate why we would prefer having `Curve3D` shown in the dataclass attributes and have the `@property` and `@Curve3DValue.setter` for the `Curve3DValues` attribute.

That being said, from a practical stand point it doesn't change anything for users. The Brep class will still expose `Curve3D` and `Curve3DValues` as public attributes (ie: attributes that don't start with `_`). If we were still using `pydantic` I would propose making `Curve3DValues` a private attribute like `_Curve3DValues` and use the `ModelField` `alias` attribute to ensure that (de)serialization occurs as expected.

Regardless, I'll change them just now as it makes sense from a style/linting perspective. It's worth bearing in mind though that the `Edges`, `Loops` and `Faces` will still have to be set using the `getter` and `setter` method so we can inject the `_brep` into them. Or maybe now is a good time to ask why we need to inject the `_brep` into them? It's pretty recursive and feels a bit dodgy.

Comment (AntoineDao) [reply specklepy/objects/geometry.py:641]: I resolved this in 5bc6b8c. Let me know what you think :raised_hands:

**Comment by AntoineDao** [comment]
Bump for attention :smile:

**Comment by izzylys** [comment]
hey @AntoineDao! sorry about the slow response on this;l we were enjoying Tuscany last week (see [Speckle Team Retreat](https://speckle.systems/blog/speckle-team-retreat/)) and I hadn't gotten a chance to revisit this yet this week. I'll have a look this afternoon!

**Comment by AntoineDao** [comment]
1. Jealous! :angry: 
2. No worries and no rush. I figured I'd just bump you for attention in case it had flown off your radar :raised_hands:

**Review by izzylys — APPROVED** [review]

**Comment by izzylys** [comment]
got a chance to dig into this and everything looks grand! behaviour seems to emulate core perfectly and you addressed every concern I had 🎉

thank you sooo much for your fab work on this antoine! the pile of 🍻 owed to you grows even larger 😅

**Comment by izzylys** [comment]
v2.3.1 deployed 🎉
